### ISSUE TITLE
Configure webpack 2 to use tree-shaking for smallest bundle size

### ISSUE THREAD

**Issue by kumar303** [mozilla/addons-frontend#2407]
Now that we're on webpack 2 we can optimize the bundle by removing unused code via tree-shaking.

Webpack2 supports es modules directly so compiling to commonjs is no longer necessary.

See:
* https://webpack.js.org/guides/tree-shaking/
* https://blog.craftlab.hu/how-to-do-proper-tree-shaking-in-webpack-2-e27852af8b21

**Comment by muffinresearch** [comment]
Worth noting from https://facebook.github.io/jest/docs/en/webpack.html#using-with-webpack-2

> webpack 2 offers native support for ES modules. However, Jest runs in Node, and thus requires ES modules to be transpiled to CommonJS modules. As such, if you are using webpack 2, you most likely will want to configure Babel to transpile ES modules to CommonJS modules only in the test environment.

**Comment by willdurand** [comment]
We can probably do this now that we have babel 7, webpack 4 and a cleaner config.

**Comment by willdurand** [comment]
Looks like there are issues with our current codebase, which prevents tree-shaking to be applied. For example, having `__filename` in our code is a problem (but we need it to generate unique IDs between the client and the server).

**Comment by willdurand** [comment]
So, it looks like we have dead code elimination enabled and it works well. Tree-shaking is not enabled yet, though but it seems to be tricky to enable for some reasons.

The main reason is: CSS order is not respected in the final bundle, which causes issue. Another issue is that some polyfill functions get removed (like `regeneratorRuntime`) but I was able to bring them back. Everything is related to the `sideEffects` config parameter.

My last experiments show that with tree-shaking enabled (assuming CSS order gets fixed), the gain would be < 10KB.

**Comment by willdurand** [comment]
It is also worth mentioning that in most docs, it is mentioned that we should set `modules: false` in `babel-preset-env`'s options. Yet, with Babel 7 (which we use), this is not needed anymore because the default value is `auto`, and in the context of Webpack, it does not generate CommonJS modules (i.e. it keeps the `import`/`export` statements, so that tree-shaking can work).

I have noticed NO difference with or without `modules: false`.

**Comment by willdurand** [comment]
So. I thought I could setup tree shaking on our apps, but that's not working as intended for many reasons. I am re-stating what I wrote in #7042 below.

TL;DR: fully enabling tree shaking on our apps will not provide any immediate benefits.

I updated our config to enable tree shaking. I thought I had to change the way `babel-preset-env` was configured but with Babel 7, it is automatically configured and "tree-shaking"-ready. We do not have to explicitly set `modules: false`, because that's already done when Babel detects that compilation happens in the context of webpack.

The next thing to configure is `sideEffects`. This is a new property to add to the `package.json` file and/or the webpack rules for each file type. This is required because otherwise you'll shake the tree way too much 😅 In other words, if you do not flag a file as side-effectful and this file is not used, webpack will remove it. The problem is that SCSS files are imported but never used in our code (!!), so without the `sideEffects` configuration, webpack will remove the SCSS files, and the (S|CSS) loaders won't extract any style anymore (the final CSS bundle will be mostly empty). This applies for all non-JS files, basically. That has been configured in #7042.

Yet, some JS files are also side-effectful. Polyfills for example. Most polyfills modify the `global` context, which means there is no immediate use in our source code, but our code won't work without them. The most obvious error I got was related to our saga layer (with the generator and async functions). Most (if not all) polyfills were removed and `regeneratorRuntime` was thus undefined. 😥

In #7042, `src/core/polyfills.js` has been marked as side-effectful too.

At this point, tree shaking was correctly configured and functional. The JS code compiled successfully, but the final CSS file was not correctly built. There is a bug that changes the order of the final CSS bundle, which creates UI issues. I cannot find why we have this bug though, but there are plenty of issues related to "webpack and wrong CSS order" 🤦‍♂️ While for most CSS code, it is OK to have a random order, `normalize.css` and the `font-face` definitions should appear first. With tree-shaking enabled, both appear in the middle of the final CSS bundle...

Last but not least, tree-shaking results are not super impressive. Numbers below.

### AMO

without tree shaking:

```
 PASS  ./dist/amo-9a09628a441f60638413.js: 410.5KB < maxSize 430KB (gzip)
```

with tree shaking:

```
 PASS  ./dist/amo-ee2a425f7f0df076aa31.js: 410.6KB < maxSize 430KB (gzip)
```

In both cases, the final JS bundle size is the same. Sometimes it is even slightly bigger when tree-shaking is not enabled. That's not a surprise and quite a few people had similar results. That being said, I consider those results equivalent and I don't think the second size is actually bigger than the first one.

### Disco

without tree shaking:

```
 PASS  ./dist/disco-6930685d87f49bdfad61.js: 240.84KB < maxSize 430KB (gzip)
```

with tree shaking and polyfills:

```
 PASS  ./dist/disco-bfe300d9524e39d29a6e.js: 240.22KB < maxSize 430KB (gzip)
```

Here is the most impressive gain: `0.62KB`

FTR, with tree shaking and no polyfills, that would be a bit better (but that breaks the app, heh):

```
 PASS  ./dist/disco-bcb455f8e4425833966d.js: 213.22KB < maxSize 430KB (gzip)
```

### Conclusion

On this basis, I am closing #7042 and this issue as well because there is nothing we can do. I don't think we should spend more time on this right now, because there is no real value IMO.

It is worth noting that dead code elimination works already (we can see this in webpack's output, by searching for "only some exports used") and that works pretty well. I guess that's why we don't have a huge gain with tree-shaking, because tree-shaking only helps finding more dead code, but I believe we only import code that we actually use, so... tree-shaking has very little value here.

**Comment by muffinresearch** [comment]
@willdurand thanks for writing this up 👍
